### PR TITLE
replace flake8-eradicate with ruff ... and enable a ton of more lint rules, fixing a couple of them.

### PR DESCRIPTION
**Pull request by jakkdl**
If enabling auto-updating of pre-commit, flake8-eradicate will come to bite us as it's not accepting PR's to show that it works with flake8 6. I was planning to add an updated fork as additional_dependency to flake8 instead, but I stumbled upon https://github.com/charliermarsh/ruff instead - which, among others, reimplements flake8-eradicate. Curiosity peaked, I added it to CI, enabled *everything* (it sucks there's no `--select-all`), and after playing around for a bit it looks to be almost strictly better than flake8 in a lot of ways. It implements a ton of plugins by default, has autofix for a bunch of them, and in general very easy to transition to from flake8.

flake8-mutable is the only plugin flake8-trio uses I didn't see listed in https://github.com/charliermarsh/ruff#rules (though maybe there is a rule for default mutable arguments anyway, haven't checked), and flake8-noqa (but there is support for checking noqa stuff).
Looks like they haven't kept up reimplementing *all* rules from flake8-bugbear (esp not from the opinionated ones), but majority of them are in.

So might be interesting to enable it, and see how viable it is as a replacement for flake8.

### CODE REVIEW

File: flake8_trio/__init__.py [hunk 1]
```diff
@@ -351,7 +351,8 @@ def parse_trio114_identifiers(raw_value: str) -> list[str]:
     values = comma_separated_list(raw_value)
     for value in values:
         if keyword.iskeyword(value) or not value.isidentifier():
-            raise ArgumentTypeError(f"{value!r} is not a valid method identifier")
+            msg = f"{value!r} is not a valid method identifier"
+            raise ArgumentTypeError(msg)
```
Comment (Zac-HD) [inline flake8_trio/__init__.py:355]: I prefer the old style here, can we disable that lint and revert?

File: flake8_trio/visitors/flake8triovisitor.py [hunk 2]
```diff
@@ -9,12 +9,12 @@
 import libcst as cst
 from libcst.metadata import PositionProvider
 
-from ..base import Error, Statement
+from flake8_trio.base import Error, Statement
```
Comment (Zac-HD) [inline flake8_trio/visitors/flake8triovisitor.py:12]: Similarly I think local imports are fine, let's disable + revert that too.

File: flake8_trio/visitors/visitor100.py [hunk 3]
```diff
@@ -7,9 +7,8 @@
 """
 from __future__ import annotations
 
-from typing import Any
+from typing import TYPE_CHECKING, Any
 
-import libcst as cst
```
Comment (Zac-HD) [inline flake8_trio/visitors/visitor100.py:12]: Aaaand, let's disable this as well - it's pretty verbose and doesn't actually have any performance benefits here at all; we're executing all the imports already when we import the matchers.

(perils of starting with such a long list of available lint rules!)

**Review by Zac-HD — COMMENTED** [review]
Nice, I've been vaguely interested in trying `ruff` for a while and this looks like a clean replacement.  IMO we could drop the flake8 checks immediately, if there's nothing else left out?

File: flake8_trio/visitors/visitor100.py (hunk 3 above)
Comment (jakkdl) [reply flake8_trio/visitors/visitor100.py:12]: Huh, looks like flake8-type-checking has explicit logic about ignoring the case when you import a submodule. I'll add `libcst` to https://beta.ruff.rs/docs/settings/#exempt-modules for now but might want to stick with `flake8-type-checking` over ruff and/or open an issue in ruff.

**Comment by jakkdl** [comment]
> IMO we could drop the flake8 checks immediately, if there's nothing else left out?

I don't mind running both for a while to see where they agree/disagree. But if it becomes bothersome I'll quickly start disabling error codes/plugins, set the [`stage`](https://pre-commit.com/#confining-hooks-to-run-at-certain-stages) for flake8 to `manual`, and/or remove it.

File: flake8_trio/visitors/visitor100.py (hunk 3 above)
Comment (jakkdl) [reply flake8_trio/visitors/visitor100.py:12]: opened https://github.com/charliermarsh/ruff/issues/4667

**Review by Zac-HD — APPROVED** [review]
:shipit:

**Pull request merged**